### PR TITLE
new plugin: `com.palantir.baseline-exact-dependencies` helps declare necessary and sufficient dependencies

### PR DESCRIPTION
**Pull request by iamdanfox**
Mostly lifted from @gatesn's original PR (https://github.com/palantir/gradle-baseline/pull/547).

This PR adds a new plugin, with two new tasks.  Neither of them run automatically as part of `./gradlew check` - expecting people to just try these out locally initially.

```
> Task :log-receiver:checkImplicitDependencies
Found 18 used but undeclared dependencies - consider adding the following explicit dependencies to 'log-receiver/build.gradle', or avoid using classes from these jars:
    dependencies {
        implementation 'com.fasterxml.jackson.core:jackson-annotations'
        implementation 'com.fasterxml.jackson.core:jackson-databind'
        implementation 'com.google.code.findbugs:findbugs-annotations'
        implementation 'com.google.code.findbugs:jsr305'
        implementation 'com.google.errorprone:error_prone_annotations'
        implementation 'com.google.guava:guava'
        implementation 'com.palantir.conjure.java.api:service-config'
        implementation 'com.palantir.conjure.java:conjure-lib'
        implementation 'com.palantir.log-receiver:log-receiver-api-objects'
        implementation 'com.palantir.safe-logging:safe-logging'
        implementation 'com.palantir.sls.logging:sls-logging'
        implementation 'com.palantir.sls.logging:sls-logging-api-objects'
        implementation 'com.palantir.tokens:auth-tokens'
        implementation 'com.palantir.tritium:tritium-registry'
        implementation 'com.palantir.witchcraft:witchcraft-logging-spi'
        implementation 'io.dropwizard.metrics:metrics-core'
        implementation 'javax.ws.rs:javax.ws.rs-api'
    }
```

```
Execution failed for task ':log-receiver:checkUnusedDependencies'.
> Found 1 extraneous dependencies, please delete them from 'log-receiver/build.gradle' or choose one of the suggested fixes:
        com.google.auto.service:auto-service
                Did you mean:
                        implementation 'com.google.errorprone:error_prone_annotations' 
                        implementation 'com.google.code.findbugs:jsr305' 
                        implementation 'com.google.guava:guava' 
```

### Known issues

- conjure-generated projects often have extraneous dependencies
- conjure-generated projects often rely on transitively provided dependencies (from conjure-lib)
- suggested fixes are not very smart (they don't suggest `project(':foo')`)
- doesn't know anything about annotation processors, or runtime-loaded jars (e.g. slf4j implementations)
- doesn't have a `--fix` variant yet.

Nonetheless, we've found some interesting stuff already!

### CODE REVIEW

**Comment by iamdanfox** [comment]
False positives we're seeing:
```
Execution failed for task ':proj:checkUnusedDependencies'.
> Declared but unused dependencies:
        auto-service.jar (com.google.auto.service:auto-service:1.0-rc4)
                Did you mean:
                        error_prone_annotations.jar (com.google.errorprone:error_prone_annotations:2.3.2)
                        jsr305.jar (com.google.code.findbugs:jsr305:3.0.2)

```

But we do actually need this dep:
```
compileOnly 'com.google.auto.service:auto-service'
```

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+        // Construct caches of classes in each dependency
+        allArtifacts.forEach(artifact -> {
+            try {
+                // Construct class/artifact maps
+                Set<String> classesInArtifact = classAnalayzer.analyze(artifact.getFile().toURI().toURL());
+                classesFromArtifact.put(artifact, classesInArtifact);
+                classesInArtifact.forEach(clazz -> {
+                    if (classToDependency.put(clazz, artifact) != null) {
+                        getLogger().info("Found duplicate class " + clazz);
```
Comment (carterkozak) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: `getLogger().info("Found duplicate class {}", clazz);`

Comment (schlosna) [reply gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: might be worth including the artifacts with duplicate classes as well

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                .collect(Collectors.toList());
+
+        List<ResolvedArtifact> declaredButUnused = Sets.difference(declaredArtifacts, expectedArtifacts).stream()
+                .sorted(Comparator.comparing(artifact -> artifact.getId().getDisplayName()))
+                .collect(Collectors.toList());
+
+        if (!usedButUndeclared.isEmpty()) {
+            getLogger().warn(
+                    "Used but undeclared dependencies:\n{}",
+                    Joiner.on("\t\n").join(usedButUndeclared));
```
Comment (schlosna) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: Do you want `"\n\t"` as separator?

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            throw new GradleException(sb.toString());
+        }
+    }
+
+    @Input
+    public List<Configuration> getDependenciesConfigurations() {
+        return dependenciesConfigurations.stream().filter(Objects::nonNull).collect(Collectors.toList());
+    }
+
+    public void setDependenciesConfigurations(List<Configuration> dependenciesConfigurations) {
```
Comment (uschi2000) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: I think gradle 5 has better ways for injecting parameters, no?

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @TaskAction
+    public void checkUnusedDependencies() {
+        Set<ResolvedDependency> declaredDependencies = getDependenciesConfigurations().stream()
+                .map(Configuration::getResolvedConfiguration)
+                .flatMap(resolved -> resolved.getFirstLevelModuleDependencies().stream())
+                .collect(Collectors.toSet());
+
+        // All artifacts
```
Comment (uschi2000) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: nit: redundant comment

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                // Construct class/artifact maps
+                Set<String> classesInArtifact = classAnalayzer.analyze(artifact.getFile().toURI().toURL());
+                classesFromArtifact.put(artifact, classesInArtifact);
+                classesInArtifact.forEach(clazz -> {
+                    if (classToDependency.put(clazz, artifact) != null) {
+                        getLogger().info("Found duplicate class " + clazz);
+                    }
+                });
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (uschi2000) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: `new RuntimeException("Failed to...", e);`

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                })
+                .collect(Collectors.toSet());
+
+        // Go through classes dirs to find which classes are referenced
+        Set<String> referencedClasses = Streams.stream(getClasses().iterator())
+                .flatMap(file -> {
+                    try {
+                        return dependencyAnalyzer.analyze(file.toURI().toURL()).stream();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
```
Comment (uschi2000) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: as above

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().withPlugin("java", plugin -> {
+            project.getTasks().create("checkExactDependencies", CheckExactDependenciesTask.class, task -> {
+                task.dependsOn(JavaPlugin.CLASSES_TASK_NAME);
+
+                SourceSetContainer sourceSets = project.getConvention()
+                        .getPlugin(JavaPluginConvention.class).getSourceSets();
+
+                SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
```
Comment (uschi2000) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java]: does this need to be configurable?

**Comment by markelliot** [comment]
it'd be pretty amazing if this had a '--fix' option

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+import org.gradle.api.tasks.TaskAction;
+
+public class CheckExactDependenciesTask extends DefaultTask {
+
+    // TODO(dfox): hoist these indexes so they are not recomputed for every subproject
+    private final Map<String, ResolvedArtifact> classToDependency = new HashMap<>();
+    private final Map<ResolvedArtifact, Set<String>> classesFromArtifact = new HashMap<>();
+    private final Map<ResolvedArtifact, ResolvedDependency> artifactsFromDependency = new HashMap<>();
+
+    private final ListProperty<Configuration> dependenciesConfigurations;
```
Comment (gatesn) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]:
```suggestion
    private final ConfigurableFileCollection dependenciesConfigurations = getProject().getLayout().configurableFiles();
```

Comment (gatesn) [reply gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: I actually meant this for the line below..

Comment (iamdanfox) [reply gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExactDependenciesTask.java]: I don't want to accidentally take a super high gradle dependency here... I think getLayout().configurableFiles() is a new 5.0 feature...

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckImplicitDependenciesTask.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        dependenciesConfigurations.set(Collections.emptyList());
+        sourceClasses = getProject().getObjects().property(FileCollection.class);
+        ignore = getProject().getObjects().setProperty(String.class);
+        ignore.set(Collections.emptySet());
+    }
+
+    @TaskAction
+    public final void checkImplicitDependencies() {
+        Set<ResolvedDependency> declaredDependencies = dependenciesConfigurations.get().stream()
+                .map(Configuration::getResolvedConfiguration)
```
Comment (ferozco) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckImplicitDependenciesTask.java:64]: Could we avoid resolvingConfigurations (and potentially compiling the whole repo) by looking at the incoming artifacts instead?

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+                // this is liberally applied to ease the Java8 -> 11 transition
+                task.ignore("javax.annotation", "javax.annotation-api");
+            });
+
+            project.getTasks().create("checkImplicitDependencies", CheckImplicitDependenciesTask.class, task -> {
+                task.dependsOn(JavaPlugin.CLASSES_TASK_NAME);
+                task.setSourceClasses(mainSourceSet.getOutput().getClassesDirs());
+                task.dependenciesConfiguration(compileClasspath);
+
+                task.ignore("org.slf4j", "slf4j-api");
```
Comment (ferozco) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java:78]: can you add a comment explaining why

**Review by ferozco — APPROVED** [review]
Overall looks good! Happy to have the changes addressed in a follow up

**Pull request merged**